appknox/reports_test.go

```diff
@@ -0,0 +1,352 @@
+
+func TestReportsService_GetReportURL_Success(t *testing.T) {
+	client, mux, _, teardown := setup()
+	defer teardown()
+
+	// Starting fake server to accept request
+	mux.HandleFunc("/api/v2/reports/1/pdf", func(w http.ResponseWriter, r *http.Request) {
+		testMethod(t, r, "GET")
+		fmt.Fprint(w, `{"url":"http://example.com"}`)
+	})
+
+	report, err := client.Reports.GetReportURL(context.Background(), 1)
+
+	if err != nil {
+		t.Errorf("Reports.GetReportURL returned error: %v", err)
+	}
+
+	want := &Report{URL: "http://example.com"}
+	if !reflect.DeepEqual(report, want) {
+		t.Errorf("Reports.GetReportURL returned %+v, want %+v", report, want)
+	}
+}
+
+func TestReportsService_GetReportURL_IFAPINotWorking_ShuldFail(t *testing.T) {
+	client, mux, _, teardown := setup()
+	defer teardown()
+
+	// Starting fake server to accept request
+	mux.HandleFunc("/api/v2/reports/1/pdf", func(w http.ResponseWriter, r *http.Request) {
+		testMethod(t, r, "GET")
+		w.WriteHeader(http.StatusInternalServerError)
+	})
+
+	_, err := client.Reports.GetReportURL(context.Background(), 1)
+
+	// Assert http response code
+	assert.Equal(t, strings.TrimSpace(strings.Split(err.Error(), ":")[3]), "500")
+}
+
+func TestReportsService_GetReportURL_IFBaseURL_ShuldFail(t *testing.T) {
+	client, mux, _, teardown := setup()
+	defer teardown()
+
+	// Starting fake server to accept request
+	mux.HandleFunc("/api/v2/reports/1/pdf", func(w http.ResponseWriter, r *http.Request) {
+		testMethod(t, r, "GET")
+		w.WriteHeader(http.StatusInternalServerError)
+	})
+
+	// Setting invalid schame name to fail in `client.NewRequest`
+	client.BaseURL.Scheme = ""
+	client.BaseURL.Host = "localhost/"
+
+	_, err := client.Reports.GetReportURL(context.Background(), 1)
+	assert.Equal(t, true, err != nil)
+}
+
+func TestReportsService_DownloadFile_Success(t *testing.T) {
+	client, mux, serverURL, teardown := setup()
+	defer teardown()
+
+	// Starting fake server to accept download request
+	mux.HandleFunc("/aws_fake_signed_url.txt", func(w http.ResponseWriter, r *http.Request) {
+		testMethod(t, r, "GET")
+		fmt.Fprint(w, `Fake_File_Content`)
+	})
+
+	outputDir := ".."
+	report, err := client.Reports.DownloadFile(context.Background(), serverURL+"/aws_fake_signed_url.txt?signature=fake_signature_hash", outputDir)
+
+	if err != nil {
+		t.Errorf("Reports.DownloadFile returned error: %v", err)
+	}
+
+	want := "../aws_fake_signed_url.txt"
+	if !reflect.DeepEqual(report, want) {
+		t.Errorf("Reports.DownloadFile returned %+v, want %+v", report, want)
+	}
+
+	// remove files after test
+	err = os.Remove(want)
+	assert.Equal(t, nil, err)
+}
+
+func TestReportsService_DownloadFile_IfWrongURL_Should_Fail(t *testing.T) {
+	client, mux, serverURL, teardown := setup()
+	defer teardown()
+
+	// Starting fake server to accept request
+	mux.HandleFunc("/aws_fake_signed_url.txt", func(w http.ResponseWriter, r *http.Request) {
+		testMethod(t, r, "GET")
+		w.WriteHeader(http.StatusNotFound)
+	})
+
+	outputDir := ".."
```

Comment:
Use `t.TempDir()` to generate a temporary directory for testing. This directory will be automatically deleted upon test completion.